File: 93-restore-ip-addresses/restore-ip-addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        ans = []

        def backtrack(idx, parts, curr):
            # If 4 parts are formed
            if parts == 4:
                if idx == len(s):
                    ans.append(".".join(curr))
                return

            # Try segments of length 1 to 3
            for length in range(1, 4):
                if idx + length > len(s):
                    break

                part = s[idx:idx + length]

                # No leading zeros
                if len(part) > 1 and part[0] == '0':
                    break

                # Value must be <= 255
                if int(part) <= 255:
                    curr.append(part)
                    backtrack(idx + length, parts + 1, curr)
                    curr.pop()

        # Quick pruning
        if len(s) < 4 or len(s) > 12:
            return []

        backtrack(0, 0, [])
        return ans

File: 93-restore-ip-addresses/restore-ip-addresses.cpp
class Solution {
public:
    vector<string> ans;

    void backtrack(string &s, int idx, int parts, string curr) {
        // If 4 parts are formed
        if (parts == 4) {
            if (idx == s.size()) {
                curr.pop_back();          // Remove last '.'
                ans.push_back(curr);
            }
            return;
        }

        // Try segments of length 1 to 3
        for (int len = 1; len <= 3 && idx + len <= s.size(); len++) {
            string part = s.substr(idx, len);

            // Leading zero is not allowed
            if (part.size() > 1 && part[0] == '0')
                break;

            int val = stoi(part);

            if (val <= 255) {
                backtrack(s, idx + len, parts + 1, curr + part + ".");
            }
        }
    }

    vector<string> restoreIpAddresses(string s) {
        if (s.size() < 4 || s.size() > 12)
            return {};

        backtrack(s, 0, 0, "");
        return ans;
    }
};